### ISSUE TITLE
Make sure all a11y radio buttons have the name attribute

### ISSUE THREAD

**Issue by zepumph** [phetsims/sun#316]
In Reactants, Products and Leftovers as well as Build an Atom there is inconsistent focus highlight with the radio buttons.

When you tab forward (tab key) to the radio buttons, it will always highlight the first radio button, and when you tab backwards to it (shift + tab keys) the focus highlight will always be around the last radio button. 

The ideal behavior would be to have the focus highlight be around the selected item, no matter how you get to the radio buttons.

@jessegreenberg and I poked around a bit, and found that AquaRadioButtons are often not being added to AquaRadioButtonGroups, so they are just li tags floating without a container. We tried adding 'ul' and 'fieldset' tags to surround the `<li>` tags, but neither worked right out of the bag. We will keep poking.

**Comment by zepumph** [comment]
@jessegreenberg and I found some seriously strange behavior. We looked into AquaRadioButtons in AtomView in BAA.  It looks like there was no name attribute on these components, which is responsible for linking them to some extent. In the chrome debugger we added the name attributes directly and all looked good (the radio button behavior was fixed, and worked like MAL which uses RadioButtonGroup). Then tried to add these two lines to the code itself in `AtomView`:

```js
orbitsRadioButton.setAccessibleAttribute( 'name', 'hello');
    cloudRadioButton.setAccessibleAttribute( 'name', 'hello');

```

And although the name attribute would show up, the keyboard navigation was completely gone.

**Comment by jessegreenberg** [comment]
BASE uses VerticalAquaRadioButtonGroup and this isn't a problem in that sim. RPAL and BAA use AquaRadioButtons without using VerticalAquaRadioButtonGroup, so we should compare the generated HTML again.

**Comment by jessegreenberg** [comment]
Accessible HTML in BAA:
```html
<li tabindex="-1" id="container-12-197-589-578">
    <label tabindex="-1" id="label-12-197-589-578" for="12-197-589-578"></label>
    <input id="12-197-589-578" tabindex="0" checked="checked" type="radio" style="width: 1px;">
</li>
<li tabindex="-1" id="container-12-197-589-587">
    <label tabindex="-1" id="label-12-197-589-587" for="12-197-589-587"></label>
    <input id="12-197-589-587" tabindex="0" type="radio" style="width: 1px;">
</li>
```

HTML in BASE:
```html
<ul tabindex="-1" id="64-125-359-435-433-403">
    <li tabindex="-1" id="container-64-125-359-435-433-403-383">
        <label tabindex="-1" id="label-64-125-359-435-433-403-383" for="64-125-359-435-433-403-383">‪Show all charges‬</label>
        <input id="64-125-359-435-433-403-383" tabindex="0" checked="checked" name="1" type="radio" style="width: 1px;">
    </li>
    <li tabindex="-1" id="container-64-125-359-435-433-403-392">
        <label tabindex="-1" id="label-64-125-359-435-433-403-392" for="64-125-359-435-433-403-392">‪Show no charges‬</label>
        <input id="64-125-359-435-433-403-392" tabindex="0" name="1" type="radio" style="width: 1px;">
    </li>
    <li tabindex="-1" id="container-64-125-359-435-433-403-401">
        <label tabindex="-1" id="label-64-125-359-435-433-403-401" for="64-125-359-435-433-403-401">‪Show charge differences‬</label>
        <input id="64-125-359-435-433-403-401" tabindex="0" name="1" type="radio" style="width: 1px;">
    </li>
</ul>
```

**Comment by jessegreenberg** [comment]
Differences I see are that BAAs buttons are not in a list, don't have the name attribute, and the labels are empty. Wondering if the browser is getting confused with an empty ```<label for="id"></label>``` element?

**Comment by jessegreenberg** [comment]
When I change the HTML in BAA to look like this (adding ul, name, and label content)

```html
<ul tabindex="-1" id="12-197-589">
    <li tabindex="-1" id="container-12-197-589-578">
        <label tabindex="-1" id="label-12-197-589-578" for="12-197-589-578">Test</label>
        <input id="12-197-589-578" tabindex="0" checked="checked" name="5" type="radio" style="width: 1px;">
    </li>
    <li tabindex="-1" id="container-12-197-589-587">
        <label tabindex="-1" id="label-12-197-589-587" for="12-197-589-587">test</label>
        <input id="12-197-589-587" tabindex="0" name="5" type="radio" style="width: 1px;">
    </li>
</ul>
```

Both radio buttons are removed from focus order. But when I press one of the buttons with the mouse, they are focusable and then keyboard focus  works exactly as it should.  So what happens when we click on them? I see no change in the HTML.

**Comment by pixelzoom** [comment]
@jessegreenberg said:
> BASE uses VerticalAquaRadioButtonGroup and this isn't a problem in that sim. 

In case you're considering using VerticalAquaRadioButtonGroup, see https://github.com/phetsims/sun/issues/30.   It is not the right approach to grouping UI components, and I'd prefer to not find it used in sims that I'm responsible for.

**Comment by jessegreenberg** [comment]
> In case you're considering using VerticalAquaRadioButtonGroup

We won't be using VerticalAquaRadioButtonGroup everywhere, just noticed that this bug isn't happening when using it and trying to figure out what the difference is. Keyboard nav should work the same in both cases.

**Comment by zepumph** [comment]
In a11y dev meeting today we found the problem with this. It is because single aquaRadioButtons are not having their 'name' attribute set. But RadioButtonGroups do have this happening:

`      radioButton.setAccessibleAttribute( 'name', 'radioButtonGroup' + instanceCount + 'Member' );`

We think that the above is not the best solution, since the name is sorta gibberish to screen readers, but it  at least makes the keyboard nav smooth.

Steps to continue for this issue:

- [ ] Add the name attribute to RadioButton, specified with an option.
- [ ] Go through all a11y sims and look at usages, making sure that each one has names used correctly. (make sure the get all usages; BASE is using name="1" for its radio buttons).
- [ ] Implement this same option in RadioButtonGroup rather than using the `instanceCount`


Top runner for the attribute name option is `a11yNameAttribute` where the `a11y` prefix is applied to all a11y specific options that are not managed by node, but rather a subType.

@mbarlow12 agreed to take the lead on this one. Thanks!

**Comment by jessegreenberg** [comment]
> We think that the above is not the best solution, since the name is sorta gibberish to screen readers,

+1 for making the way in which the `name` attribute is set consistent, but we should determine whether or not the `name` is read by screen readers since that could have an impact on the way this option is implemented and used.  @terracoda could you please verify whether or not the `name` attribute is read by screen readers?

**Comment by mbarlow12** [comment]
First, I discovered that RadioButton.js is only used in AquaRadioButton, so I wonder if it would be worth it to 1) remove RadioButton.js all together and fold the functionality into AquaRadioButton or 2) spread it's utilization to RadioButtonGroupMember and the like. @zepumph @jessegreenberg thoughts?

Also, I worked through the a11y sims and confirmed the name attribute in all of them. Here are a few oddities:
- area model common utilizes a MutableOptionsNode. I didn't look too deeply into it, but since we're still using the `instanceCount` I expected the numbers to be sequential. They're not; we seem to be missing #1 - 3 and 6. No idea what's happening there.
- there are some odd html structures in BAA such as `li` elements outside of lists with ungrouped radio buttons. This may not be an issue, but someone with more familiarity with sim could comment.

I'll mark it for the next dev a11y discussion.

**Comment by mbarlow12** [comment]
Another note: @jessegreenberg was unsure that screen readers and other AT use this attribute. So there is currently no option to pass to the ...Group types to name their constituent buttons, and those types still use the `instanceCount` as the unique identifier.

**Comment by terracoda** [comment]
@mbarlow12, 
> `li `elements outside of lists  
is more than odd, it is not valid code and could present issues for Assistive Technology. 

I would advise taming the wild `li` tags with a parent `ul`.

Radio buttons that are not grouped with a `ul` tag or a `radiogroup role` may not be navigable with arrow keys. If they are visually in a group, we should strive to communicate them as a group to AT users.

**Comment by terracoda** [comment]
The html `name` attribute, I don't think this has any relevant effect on accessibility. I am pretty sure it is used for programmatic identification purposes, much like the id attribute and useful to identify things for servers ([](https://developer.mozilla.org/en-US/docs/Web/HTML/Attributes)), but not for users. I think the attribute is only used on form elements for form submission purposes 
[](https://www.w3.org/TR/html52/sec-forms.html#naming-form-controls-the-name-attribute). I don't think it provides any accessible information for AT.

The accessible name for elements is provided through label elements, aria-label, or aria-labelledby.

I hope that helps the discussion.

**Comment by zepumph** [comment]
>First, I discovered that RadioButton.js is only used in AquaRadioButton,

I confirmed that AquaRadioButton is the only use of RadioButton, but I don't know if it is separate for a good reason, or just because of something vestigial. @samreid may know more about its past. 


>there are some odd html structures in BAA such as li elements outside of lists with ungrouped radio buttons. This may not be an issue, but someone with more familiarity with sim could comment.

@terracoda would know best, but that sim is not in active a11y development, and so none of that mess (largely that I left) will be published anytime soon. I personally wouldn't spend a lot of time on it at all, as there is a good change that that html will all change.

**Comment by samreid** [comment]
> First, I discovered that RadioButton.js is only used in AquaRadioButton,

Good catch @mbarlow12, I'm inclined to combine them into AquaRadioButton and delete RadioButton, but we should check with @pixelzoom since he is the original author of both.

**Comment by terracoda** [comment]
Regarding my own comment (https://github.com/phetsims/sun/issues/316#issuecomment-370110686):
> Radio buttons that are not grouped with a `ul` tag or a `radiogroup` role may not be navigable with arrow keys. 

I haven't tested this, so I do not know for sure how groups of radio buttons that are not explicitly grouped with code behave.

**Comment by terracoda** [comment]
@zepumph, I agree that the lists in BAA is not a good example to be looking at because it is not in active development.

**Comment by pixelzoom** [comment]
@samreid said:
>  I'm inclined to combine them into AquaRadioButton and delete RadioButton, but we should check with @pixelzoom since he is the original author of both.

Long ago, in a galaxy far far away, the look-&-feel of HTML5 sims was very much in flux. RadioButton was intended to be the base type for all radio buttons, and AquaRadioButton was a radio button with the Aqua (macOS) look-&-feel.  Apparently AquaRadioButton was sufficient.  

My recommendation would be to fold RadioButton into AquaRadioButton, delete RadioButton, and then optionally:

- rename AquaRadioButton to RadioButton
- rename AquaRadioButtonIO to RadioButtonIO
- rename VerticalAquaRadioButtonGroup to VerticalRadioButtonGroup
- search for sim-specific types with "Aqua" in the name, remove the "Aqua"

**Comment by samreid** [comment]
What about keeping the name `AquaRadioButton` or something similar, so as not to confuse `RadioButton` with `RadioButtonGroup`?

**Comment by pixelzoom** [comment]
Fine with me, that's why I said "optional".

**Comment by terracoda** [comment]
I think I answered the a11y questions.
Regarding BAA code, I agree with @zepumph that that sim should not be used as an example. It is not in active development.

I am unassigning myself. Please re-assign if you need anything else from me.

**Comment by zepumph** [comment]
I think this issue is largely back to you @mbarlow12.

**Comment by mbarlow12** [comment]
RadioButton and RadioButtonIO removed in above commits. @zepumph or @jessegreenberg would either of you like to review?

**Comment by pixelzoom** [comment]
AquaRadioButton line 137: `this.disposeRadioButton` should be `this.disposeAquaRadioButton`

**Comment by zepumph** [comment]
I'll take a look

**Comment by zepumph** [comment]
This looks very nice to me! Thanks for doing this, back to you @mbarlow12 to review my commits and close.

**Comment by mbarlow12** [comment]
Everything looks good to me @zepumph. Closing.
